File: Java-CodeServer/src/main/resources/js/prompto/type/DecimalType.js
var NativeType = require("./NativeType").NativeType;
var BooleanType = require("./BooleanType").BooleanType;
var IntegerType = null; // circular dependency
var AnyType = require("./AnyType").AnyType;
var Decimal = require("../value/Decimal").Decimal;

exports.resolve = function() {
	IntegerType = require("./IntegerType").IntegerType;
}

function DecimalType()  {
	NativeType.call(this, "Decimal");
	return this;
}

DecimalType.prototype = Object.create(NativeType.prototype);
DecimalType.prototype.constructor = DecimalType;

DecimalType.instance = new DecimalType();


DecimalType.prototype.isAssignableTo = function(context, other) {
	return (other instanceof IntegerType) || (other instanceof DecimalType) || (other instanceof AnyType);
};


DecimalType.prototype.checkAdd = function(context, other, tryReverse) {
	if(other instanceof IntegerType) {
		return this;
	} else if(other instanceof DecimalType) {
		return this;
	} else {
		return NativeType.prototype.checkAdd.call(this, context, other, tryReverse);
	}
};

DecimalType.prototype.checkSubstract = function(context, other) {
	if(other instanceof IntegerType || other instanceof DecimalType) {
		return this;
	} else {
		return NativeType.prototype.checkSubstract.call(this, context, other);
	}
};

DecimalType.prototype.checkMultiply = function(context, other, tryReverse) {
	if(other instanceof IntegerType || other instanceof DecimalType) {
		return this;
	} else {
		return NativeType.prototype.checkMultiply.call(this, context, other, tryReverse);
	}
};

DecimalType.prototype.checkDivide = function(context, other) {
	if(other instanceof IntegerType || other instanceof DecimalType) {
		return this;
	} else {
		return NativeType.prototype.checkDivide.call(this, context, other);
	}
};

DecimalType.prototype.checkIntDivide = function(context, other) {
    if(other instanceof IntegerType) {
        return this;
    } else {
        return NativeType.prototype.checkIntDivide.call(this, context, other);
    }
};

DecimalType.prototype.checkModulo = function(context, other) {
    if(other instanceof IntegerType || other instanceof DecimalType) {
        return this;
    } else {
        return NativeType.prototype.checkModulo.call(this, context, other);
    }
};

DecimalType.prototype.checkMinus = function(context) {
	return this;
};

DecimalType.prototype.checkCompare = function(context, other) {
	if(other instanceof IntegerType || other instanceof DecimalType) {
		return BooleanType.instance;
	} else {
		return NativeType.prototype.checkCompare.call(this, context, other);
	}
};

DecimalType.prototype.convertJavaScriptValueToPrestoValue = function(context, value, returnType) {
    if (typeof(value)=='number') {
        return new Decimal(value);
    } else {
        return value; // TODO for now
    }
};

exports.DecimalType = DecimalType;

File: Java-Core/src/main/java/prompto/type/DecimalType.java
package prompto.type;

import java.lang.reflect.Type;
import java.util.Comparator;

import prompto.parser.ISection;
import prompto.runtime.Context;
import prompto.value.Decimal;
import prompto.value.IValue;

import com.fasterxml.jackson.databind.JsonNode;


public class DecimalType extends NativeType implements INumberType {

	static DecimalType instance = new DecimalType();
	
	public static DecimalType instance() {
		return instance;
	}
	
	private DecimalType() {
		super(Family.DECIMAL);
	}
	
	@Override
	public Type getJavaType(Context context) {
		return Double.class;
	}
	

	@Override
	public boolean isAssignableTo(Context context, IType other) {
		return (other instanceof IntegerType) || (other instanceof DecimalType) || (other instanceof AnyType);
	}
	
	@Override
	public IType checkAdd(Context context, IType other, boolean tryReverse) {
		if(other instanceof IntegerType)
			return this;
		if(other instanceof DecimalType)
			return this;
		return super.checkAdd(context, other, tryReverse);
	}
	
	@Override
	public IType checkSubstract(Context context, IType other) {
		if(other instanceof IntegerType)
			return this;
		if(other instanceof DecimalType)
			return this;
		return super.checkSubstract(context, other);
	}

	@Override
	public IType checkMultiply(Context context, IType other, boolean tryReverse) {
		if(other instanceof IntegerType)
			return this;
		if(other instanceof DecimalType)
			return this;
		return super.checkMultiply(context, other, tryReverse);
	}

	@Override
	public IType checkDivide(Context context, IType other) {
		if(other instanceof IntegerType)
			return this;
		if(other instanceof DecimalType)
			return this;
		return super.checkDivide(context, other);
	}
	
	@Override
	public IType checkIntDivide(Context context, IType other) {
		if(other instanceof IntegerType)
			return other;
		return super.checkIntDivide(context, other);
	}
	
	@Override
	public IType checkModulo(Context context, IType other) {
		if(other instanceof IntegerType)
			return this;
		if(other instanceof DecimalType)
			return this;
		return super.checkModulo(context, other);
	}

	@Override
	public IType checkCompare(Context context, IType other, ISection section) {
		if(other instanceof IntegerType)
			return BooleanType.instance();
		if(other instanceof DecimalType)
			return BooleanType.instance();
		return super.checkCompare(context, other, section);
	}

	@Override
	public Comparator<Decimal> getComparator() {
		return new Comparator<Decimal>() {
			@Override
			public int compare(Decimal o1, Decimal o2) {
				return java.lang.Double.compare(o1.doubleValue(), o2.doubleValue());
			}
		};
	}

	@Override
	public IValue convertJavaValueToPromptoValue(Context context, Object value) {
        if (value instanceof Number)
            return new Decimal(((Number)value).doubleValue());
        else
            return (IValue)value; // TODO for now
	}
	
	@Override
	public IValue readJSONValue(Context context, JsonNode value) {
		return new Decimal(value.asDouble());
	}

}
